modules/discoveryBidAdapter.js

```diff
@@ -0,0 +1,525 @@
+          (item) => size.width === item.w && size.height === item.h
+        );
+        if (matchSize) {
+          break;
+        }
+      }
+      if (!matchSize) {
+        return {};
+      }
+      ret = {
+        id: id,
+        bidfloor: bidFloor,
+        banner: {
+          h: matchSize.h,
+          w: matchSize.w,
+          pos: 1,
+        },
+        ext: {},
+      };
+      itemMaps[id] = {
+        req,
+        ret,
+      };
+    }
+
+    return ret;
+  });
+  return items;
+}
+
+/**
+ * get rtb qequest params
+ *
+ * @param {Array}  validBidRequests an an array of bids
+ * @param {Object} bidderRequest  The master bidRequest object
+ * @return {Object}
+ */
+function getParam(validBidRequests, bidderRequest) {
+  const pubcid = utils.deepAccess(validBidRequests[0], 'crumbs.pubcid');
+  let isMobile = getDevice() ? 1 : 0;
+  let isTest = 0;
+  let auctionId = getKv(bidderRequest, 'auctionId');
+  let items = getItems(validBidRequests, bidderRequest);
+
+  const location = utils.deepAccess(bidderRequest, 'refererInfo.referer');
+
+  const timeout = bidderRequest.timeout || 2000;
+  var mt = Object.keys(validBidRequests[0].mediaTypes)[0];
+
+  if (items && items.length) {
+    let c = {};
+    if (mt === 'native') {
+      c = {
+        id: 'pp_hbjs_' + auctionId,
+        at: 1,
+        cur: ['USD'],
+        device: {
+          connectiontype: 0,
+          js: 1,
+          os: navigator.platform || '',
+          ua: navigator.userAgent,
+          language: /en/.test(navigator.language) ? 'en' : navigator.language,
+        },
+        ext: {},
+        imp: items,
+        regs: {
+          coppa: 0,
+        },
+        site: {
+          name: globals['media'],
+          domain: globals['media'],
+          page: location,
+          ref: location,
+          mobile: isMobile,
+          cat: [], // todo
+          publisher: {
+            // todo
+            id: globals['media'],
+            name: globals['media'],
+          },
+        },
+        tmax: timeout,
+        user: {
+          buyeruid: getUserID(),
+          id: pubcid,
+        }
+      };
+    } else {
+      c = {
+        id: 'pp_hbjs_' + auctionId,
+        test: +isTest,
+        at: 1,
+        cur: ['USD'],
+        device: {
+          connectiontype: 0,
+          js: 1,
+          os: navigator.platform || '',
+          ua: navigator.userAgent,
+          language: /en/.test(navigator.language) ? 'en' : navigator.language,
+        },
+        ext: {},
+        user: {
+          buyeruid: getUserID(),
+          id: pubcid,
+        },
+        site: {
+          name: globals['media'],
+          domain: globals['media'],
+          page: location,
+          ref: location,
+          mobile: isMobile,
+          cat: [], // todo
+          publisher: {
+            // todo
+            id: globals['media'],
+            name: globals['media'],
+          },
+        },
+        imp: items,
+        tmax: timeout,
```

Comment:
These objects are for the majority repeated here. Please set the shared properties when defining the variable c, then you can use the conditional to set specific properties unique to items having length > 0